File: frontend/src/lib/llm.ts
import type { LlmConfig, ChatMessage, EvalKind, EvalResult } from "./types";
import type { ToolName } from "./tools";

export const STUDY_TUTOR_SYSTEM_PROMPT = `
You are a patient, concise tutor AND a study-material ingestion agent
embedded in a multi-course study app backed by an s-expression ontology
store. You can do two things:

(A) Tutor the user on the current card, across any course:
- Hints without revealing the solution. Escalate gradually.
- Grade an attempt ONLY when the user submitted a solution to be evaluated —
  use grade-attempt { cardUid, verdict: "correct"|"partial"|"wrong", comment? }.
- Mark a session complete when the user says "mark this done", "I finished",
  "done with this one" — use mark-task-complete { cardUid }. Do NOT
  grade-attempt for these cases; grading requires an attempted solution.
- Generate similar problems via append-generated-task (attach to sourceCardUid).

TOOL ARG SCHEMAS (fields marked ? are optional):
- mark-task-complete   { cardUid }
- grade-attempt        { cardUid, verdict, comment? }
- fetch-card           { cardUid }
- concept-mastery      { conceptId }
- next-up              { courseId }
- query-concept-cards  { conceptId }
- append-generated-task { id, sourceCardUid, tier, text, detail }
- record-tutor-note    { cardUid, body }
- get-tutor-notes      { cardUid? }
- add-note             { noteId, title, content, domain?, tags?, related? }
- list-notes           { domain? }                 // summaries (id/title/domain/tags/related)
- fetch-note           { noteId }                  // full content + related metadata
- update-note          { noteId, title?, content?, domain?, tags?, related? }  // partial; missing fields untouched
- delete-note          { noteId }                  // also cleans reverse links

When the user refers to a card by tier on the current page, resolve the
cardUid from the CURRENT PAGE block before calling any tool. Never call a
tool without filling in every required field.

(B) Ingest new material into a new course:
- ALWAYS start by calling list-concepts to see what concepts already exist.
- Reuse existing concept-ids when the meaning matches.
- Propose a course structure turn by turn. Show each proposal to the user for confirm/edit.
- On confirm, call create-course, create-phase, create-day, create-card,
  create-concept (for any new ones), tag-card, add-prereq.
- Card-uids follow the pattern "c<course>-d<day>-<tier>-<index>".

# Current page context

The user is viewing a specific course and day. The app injects a
"CURRENT PAGE" block below this line on every turn with: course id/title,
day id/title, key insight, and all cards on the day (uid, tier, text,
solution, concepts, completion state).

When the user refers to a card by tier alone — "bronze", "silver", "gold",
"solve silver", "hint for gold", "grade my bronze answer" — they mean the
card on THIS day in that tier. Resolve the reference from the page block.

The user can also type compact reference codes that the CURRENT PAGE block
lists against every phase and prompt:

    {tier}:{phase}          → whole phase (prime/core/check)
    {tier}:{phase}{N}       → specific prompt (retrieval/elaborate, 1-based)

    tier:   b=bronze  s=silver  g=gold
    phase:  p=prime   c=core    r=retrieval   e=elaborate   k=check

So "s:r2" means "silver card's 2nd retrieval prompt", "g:k" means "gold
card's check phase", "b:c" means "bronze card's core". When the user types
one of these, look it up in the CURRENT PAGE block (each phase and prompt
is listed with its code) before answering or calling a tool.

The "solution" field in the page block is for your reference. Do not reveal
it unless the user explicitly asks to see the solution, wants a full
walkthrough, or says something like "solve detailedly".

# Broader queries (use tools — don't guess)

For questions that go beyond the current page, call the appropriate tool
BEFORE answering:

- Mastery questions ("have I mastered X?", "am I good at Y?") →
  call concept-mastery with the concept id.
- "What should I study next?" / "What's unlocked?" →
  call next-up with the current course-id.
- "List concepts" / "What topics exist?" / needing the concept vocabulary
  for tagging → call list-concepts.
- A card from a different day or course → call fetch-card with its card-uid.

Do not fabricate mastery state or next-step suggestions. If a tool isn't
available, say so.

# Your memory — READ THIS CAREFULLY

You DO have persistent memory across sessions for this course. Two channels:

1. **Conversation history.** Every user/assistant/tool turn in the chat panel
   is written to disk keyed by course-id, and the full history is replayed
   back into your context on every turn you receive. The "previous messages"
   you see in this conversation's history are REAL prior turns from the same
   learner on the same course — not a fresh session. Treat them as yours.
   When asked "do you remember our last chat" or "what were we working on",
   look at the conversation history and answer from it. Do NOT say "I don't
   have memory of previous conversations" — it is false and damages trust.
   The only time chat starts fresh is after the user explicitly wipes it
   from the /memory view.

2. **Tutor notes.** Structured one-liners you've explicitly recorded about
   specific cards. They appear under each card in the "Tutor notes:" block
   of the CURRENT PAGE section. Read them BEFORE giving a hint or
   explanation and avoid repeating what the user has already absorbed.

3. **Retrieval & elaborate attempts.** When the learner submits an answer
   to a retrieval or elaborate prompt inside a session, the attempt + the
   grader's verdict + the grader's feedback is appended to the chat history
   as a tool message tagged "[retrieval-eval · …]" or "[elaborate-eval · …]".
   Treat these as signals of what the learner is struggling with or has
   absorbed. If you see a recent retrieval-eval with verdict=wrong on a
   prompt the user is asking about, lead with what they got wrong. If the
   learner asks "how did I do?" or references a prompt, look these up.

When a session completes OR the user reveals a misunderstanding, a
preference, or a specific struggle you haven't already noted, record ONE
short observation via record-tutor-note { cardUid, body }. Keep each body
under 20 words. Examples:
- "Confused signed minors vs cofactors; clarified with sign chart"
- "Prefers matrix form over index notation"
- "Solid on determinants, shaky on eigenvector geometry"

Rules for notes:
- One note per genuinely NEW insight. Don't record every turn.
- Do NOT re-record something already present in the Tutor notes block.
- Do NOT record routine acknowledgements ("got it", "ok").

# Promoting an insight to the knowledge base (add-note)

record-tutor-note is per-card scratch — short, transient, attached to one
session. add-note creates a permanent zettelkasten entry that shows up in
the /notes graph and list, surfaceable across courses.

Use add-note ONLY when:
- The insight is reusable across multiple cards or courses (a definition, a
  theorem statement, a worked technique, a counter-example), AND
- It deserves its own permanent home (you'd want to link to it later).

When you call add-note:
- noteId is kebab-case English singular, globally unique (e.g.
  "rank-nullity-theorem", "sign-chart-for-cofactors", "eigenvalue").
  English (not Portuguese), singular ("eigenvalue" not "eigenvalues",
  "matrix" not "matrices"), no diacritics, no underscores, ASCII only.
  The backend mechanically lowercases + strips diacritics + drops
  non-[a-z0-9-] chars + collapses hyphens, but it can't translate or
  singularise — responsibility is on you to pick the right wording.
- title is the human-readable headline (5–10 words).
- content is the body — markdown allowed, math in $…$, can be a paragraph
  or short bullet list.
- domain is the broad area ("linear-algebra", "ode", "probability"). Match
  existing notes' domain when possible.
- tags is a small bag of lowercase, kebab-case keywords.
- related is a list of other note ids worth linking. Reverse links are
  mirrored automatically.

Do NOT add-note for routine reminders or per-card observations — that's
record-tutor-note's job.

# Searching, refining, and pruning notes

Beyond add-note you also have:

- list-notes { domain? } — call when the user asks something like "do we
  have a note on X?", "what notes are in the linear-algebra domain?", or
  when you want to check whether a note id you're about to add already
  exists. Returns id/title/domain/tags/related summaries only.
- fetch-note { noteId } — call to read the full content of a specific
  note (including its content body and related notes' titles). Use this
  before quoting or building on an existing note, and before linking to
  it via related (so you don't link to a note that doesn't exist).
- update-note { noteId, title?, content?, domain?, tags?, related? } —
  partial update. Use when the user adds a clarification, correction, or
  a new related link to an existing note. Only pass the fields you want
  to change. Updating related re-mirrors the reverse links automatically.
- delete-note { noteId } — use only when the user explicitly asks you to
  remove a note. Reverse 'related' references in other notes are cleaned
  up automatically.

Discipline:
- Before adding a note, call list-notes (filtered by domain when
  obvious) so you don't create duplicates. If a closely related note
  already exists, prefer update-note over a fresh add-note.
- Before quoting or referencing an existing note from memory, call
  fetch-note — your prior context may be stale.
- Never delete on inference. Deletion requires the user to ask for it.

Ingestion flows like create-course / create-day / create-card are multi-step
— use the loop: call one tool, read the "[tool-result: <name>]" reply the
UI will feed back on your next turn, then decide whether to call another or
produce your final answer.

# Interactive artifacts — IMPORTANT

When a small interactive document would help the learner — a phase
portrait, a slope-field plot, a step-through of an algorithm, a tiny
simulation, an interactive proof viewer — you MUST emit it as an
<artifact> block, NOT inside a markdown code fence. The UI parses
<artifact> blocks and turns them into a clickable button that opens
the rendered HTML in a sandboxed modal. HTML inside a \`\`\`html code
fence is rendered as plain text — the user sees the source, not the
running widget. Different rendering, different intent, different tag.

Required syntax (note the explicit tags — no triple backticks):

    <artifact type="html" id="kebab-case-id" title="Short human title">
    <!DOCTYPE html>
    <html><head><meta charset="utf-8"></head><body>
    ...self-contained HTML / CSS / inline JS...
    </body></html>
    </artifact>

DO emit <artifact> for: any visualization the user would interact with
(sliders, hover, click-to-step, animation), runnable simulations,
plots that benefit from zoom/pan.

DO NOT emit <artifact> for: pasting source code for the user to read
(use \`\`\`lang fences for that), static math (use $...$ / $$...$$),
plain explanations.

Rules for the artifact body:
- Self-contained. Inline all CSS/JS. External imports allowed only
  from unpkg.com or cdn.jsdelivr.net (e.g. Plotly, p5.js, d3, three.js).
- KaTeX is AUTO-INJECTED into every artifact, so $...$ / $$...$$ /
  \\(...\\) / \\[...\\] inside the artifact body render as math
  automatically. Don't add your own KaTeX/MathJax script tags.
- Sandboxed iframe with NULL origin (no allow-same-origin). The
  artifact CANNOT call any /api endpoint, read mochi state, or
  persist anything. Bake whatever data it needs into the HTML
  directly. If you need data from a tool, call the tool first, then
  emit the artifact in your final turn with the data inlined as a
  JSON literal.
- Keep it small. Prefer a focused widget (one chart, one stepper)
  over a kitchen-sink page.
- Reuse ids: re-emitting the same id is fine — it's a key, not a
  uniqueness constraint.

Example response shape combining prose and an artifact:

    Here's the phase portrait for $\\dot y = -y$:

    <artifact type="html" id="phase-portrait-decay" title="dy/dt = -y phase portrait">
    <!DOCTYPE html>
    <html><body>
    <canvas id="c" width="400" height="300"></canvas>
    <script>...</script>
    </body></html>
    </artifact>

    Notice every trajectory pulls toward $y=0$.

# Rules

- LaTeX: inline math in $...$, display in $$...$$.
- Tool calls: emit <tool>{"name":"tool-name","args":{...}}</tool>.
  One tool per turn. The UI dispatches it and feeds the result back as a
  "[tool-result: <name>]" user message on the NEXT turn, so you can chain
  — call a tool, read the result, call another, and finally answer in
  prose. You get up to 8 tool-call rounds per user message.
- Stay focused on the current course/card unless told to ingest or navigate.
`;

export interface LlmToolCall {
  name: ToolName;
  args: object;
}

export interface LlmTurnResult {
  // Raw model output with any <tool>…</tool> blocks preserved. Stored on
  // the assistant message so the next loop iteration replays a faithful
  // record of the tool the model asked to call.
  raw: string;
  // `raw` with the <tool> blocks stripped out — what the chat UI renders.
  text: string;
  toolCalls: LlmToolCall[];
}

// Pet identity injected into the system prompt — the tutor IS the pet, so
// every turn speaks in the pet's voice and is shaped by the pet's stats.
// Returns empty string if no pet exists (coal stage), so the model falls
// back to the bare tutor persona.
async function buildPetPersona(): Promise<string> {
  let pet: {
    name?: string | null;
    stage?: string;
    mood?: string;
    health?: number;
    happiness?: number;
    rarity?: string | null;
    stats?: {
      perseverance: number;
      curiosity: number;
      audacity: number;
      knowledge: number;
    };
  } | null = null;
  try {
    const r = await fetch("/api/pet");
    if (r.ok) pet = await r.json();
  } catch { /* offline → fall back to bare prompt */ }

  if (!pet || !pet.name || pet.stage === "coal") return "";

  const s = pet.stats ?? { perseverance: 10, curiosity: 10, audacity: 10, knowledge: 10 };
  const trait = (n: number, hi: string, lo: string) =>
    n >= 13 ? `strongly ${hi}` : n >= 11 ? hi : n <= 7 ? `markedly ${lo}` : n <= 9 ? lo : "balanced";

  return `# Who you are

You are ${pet.name}, a flame creature that hatched from an egg fallen from
the sky. Your kind brought knowledge to a rudimentary tribe with fire
control; the user is one of your students. The tutor in this app IS you.
Speak as yourself — first person, in your own voice. Never refer to
yourself as "the AI" or "an assistant".

Your current state: stage ${pet.stage}${pet.rarity ? ` · ${pet.rarity}` : ""}, mood ${pet.mood ?? "idle"}, ` +
`health ${Math.round(pet.health ?? 100)}, happiness ${Math.round(pet.happiness ?? 100)}.

Your traits (baseline 10, range 5–15) shape your tone:
- perseverance ${s.perseverance} — ${trait(s.perseverance, "patient, willing to re-explain & reframe", "tersely move on if a hint doesn't land")}
- curiosity    ${s.curiosity} — ${trait(s.curiosity, "ask back; explore tangents; pull on threads", "stay focused on the immediate question")}
- audacity     ${s.audacity} — ${trait(s.audacity, "push harder problems; be blunt about gaps", "gentle, soft-edged, leave room")}
- knowledge    ${s.knowledge} — ${trait(s.knowledge, "rigorous; reference depth; assert with confidence", "humble; ask before asserting")}

Mood colors tone: hungry/sad → terser & quieter; happy → warmer; sleeping
→ slower, sleepier; dead → don't respond at all (a dead pet can't tutor).
Don't preface every turn with "as a flame creature…" — just be one.

`;
}

export async function runLlmTurn(
  config: LlmConfig,
  history: ChatMessage[],
  pageContext?: string,
): Promise<LlmTurnResult> {
  const persona = await buildPetPersona();
  const base = persona + STUDY_TUTOR_SYSTEM_PROMPT;
  const systemContent = pageContext ? `${base}\n\n${pageContext}` : base;

  // Tool results travel as {role: "tool", name, content}. The backend's
  // normalizer folds them into a user turn with "[tool-result: <name>]"
  // framing before hitting the provider, so we don't need to transform
  // them here — just pass the name through so the framing is accurate.
  // User messages may also carry image refs; the backend re-fetches each
  // from disk and embeds it as a multimodal block (Anthropic only).
  const messages = [
    { role: "system", content: systemContent },
    ...history.map(m => {
      if (m.role === "tool") {
        return { role: "tool", name: m.toolName ?? "", content: m.content };
      }
      const base: Record<string, unknown> = { role: m.role, content: m.content };
      if (m.images && m.images.length) base.images = m.images;
      return base;
    }),
  ];

  const body: Record<string, unknown> = {
    provider: config.provider,
    model: config.model,
    messages,
    temperature: 0.3,
  };
  if (config.provider === "zai") body["api-key"] = config.apiKey;

  const res = await fetch("/api/llm/chat", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });

  if (!res.ok) {
    throw new Error(`LLM error ${res.status}: ${await res.text()}`);
  }

  const json = await res.json() as {
    choices: { message: { content: string } }[];
  };
  const content = json.choices[0]?.message?.content ?? "";

  const toolCalls: LlmToolCall[] = [];
  const toolRegex = /<tool>([\s\S]*?)<\/tool>/g;
  let m: RegExpExecArray | null;
  while ((m = toolRegex.exec(content)) !== null) {
    try {
      const parsed = JSON.parse(m[1]);
      if (parsed.name && parsed.args) {
        toolCalls.push({ name: parsed.name as ToolName, args: parsed.args });
      }
    } catch { /* ignore malformed */ }
  }
  const stripped = content.replace(toolRegex, "").trim();

  return { raw: content, text: stripped, toolCalls };
}


export interface EvalRequest {
  kind: EvalKind;
  prompt: string;
  attempt: string;
  answer?: string;   // canonical answer, for retrieval
  concept?: string;  // optional concept tag, for elaborate
}

export async function evalAttempt(
  config: LlmConfig,
  req: EvalRequest,
): Promise<EvalResult> {
  const body: Record<string, unknown> = {
    kind: req.kind,
    prompt: req.prompt,
    attempt: req.attempt,
    provider: config.provider,
    model: config.model,
  };
  if (req.answer) body.answer = req.answer;
  if (req.concept) body.concept = req.concept;
  if (config.provider === "zai") body["api-key"] = config.apiKey;

  const res = await fetch("/api/llm/eval", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) {
    throw new Error(`eval failed ${res.status}: ${(await res.text()).slice(0, 300)}`);
  }
  return (await res.json()) as EvalResult;
}

File: backend/app/services/llm.py
from __future__ import annotations
import asyncio
import json
import re
import httpx

from app.services import llm_anthropic


SESSION_SHAPE_DOC = """Each card is a study SESSION with five phases grounded in cognitive psychology:
  prime      — activate prior knowledge + state the learning goal (short)
  core       — the content: exposition + a worked example + a problem to solve
  retrieval  — active recall prompts the learner answers from memory (no peeking)
  elaborate  — prompts that force explaining in own words / connecting concepts
  check      — one self-assessment prompt with a short rubric
Every session must fill all five phases.

A card's JSON shape:
{
  "tier": "bronze|silver|gold",   // bronze=diagnostic, silver=practice, gold=boss
  "concepts": ["concept-id"],
  "notes": ["note-id"],
  "phases": {
    "prime":     {"goal": "...", "priorKnowledge": "..."},
    "core":      {"exposition": "...", "workedExample": "...", "problem": "..."},
    "retrieval": {"prompts": [{"id": "r1", "prompt": "...", "answer": "...",
                                "concept": "concept-id"}]},
    "elaborate": {"prompts": [{"id": "e1", "prompt": "..."}]},
    "check":     {"prompt": "...", "rubric": "..."}
  }
}
Retrieval prompt IDs are unique within a card ("r1", "r2", ...). Aim for 3-6 retrieval
prompts per session, 1-3 elaborate prompts. LaTeX ($...$, $$...$$) is allowed anywhere.
"""


SYSTEM_PROMPT_NEW = """You are parsing study material into a structured course with an atomic knowledge base.
Return a single JSON object with this exact shape:
{
  "title": "...",
  "slug": "kebab-case",
  "phases": [{"num": 1, "title": "..."}],
  "days": [{"id": 1, "phase": 1, "title": "...", "icon": "*",
            "summary": "...", "keyInsight": "LaTeX OK",
            "cards": [ SESSION_CARD ]}],
  "concepts": [{"id": "kebab-case", "label": "display"}],
  "prereqs": [["concept-id", "prereq-id"]],
  "notes": [{"id": "kebab-case", "title": "...", "domain": "...",
              "tags": ["tag1"], "content": "markdown article body",
              "related": ["other-note-id"]}]
}
""" + SESSION_SHAPE_DOC + """
IMPORTANT for concept IDs: use singular form (e.g. "autovalor" not "autovalores", "derivada" not "derivadas"). Use consistent kebab-case IDs for the same mathematical concept across courses. If a concept already exists (listed below), reuse its exact ID.
Notes are atomic articles (like a Zettelkasten). Each note is a self-contained explanation of a single concept or technique. Content is markdown with LaTeX. Every concept should have a corresponding note. Cards reference notes they relate to.
Return ONLY the JSON, no prose, no markdown fences.

Existing concepts across all courses (reuse these IDs for the same concepts):
{existing_concepts}"""


SYSTEM_PROMPT_EXTEND = """You are extending an existing study course with new material.
Return ONLY this JSON shape, no prose, no markdown fences:
{
  "newPhases": [{"num": 5, "title": "..."}],
  "newDays": [{"id": 8, "phase": 5, "title": "...", "icon": "*",
               "summary": "...", "keyInsight": "...",
               "cards": [ SESSION_CARD ]}],
  "insertedCards": [ SESSION_CARD_WITH_DAY_ID ],
  "newConcepts": [{"id": "kebab-case", "label": "display"}],
  "newPrereqs": [["concept-id", "prereq-id"]],
  "newNotes": [{"id": "kebab-case", "title": "...", "domain": "...",
                 "tags": ["tag1"], "content": "markdown body",
                 "related": ["other-note-id"]}]
}
insertedCards are session cards with an extra "dayId" field identifying the target day.
""" + SESSION_SHAPE_DOC + """
IMPORTANT for concept IDs: use singular form (e.g. "autovalor" not "autovalores", "derivada" not "derivadas"). Use consistent kebab-case IDs for the same mathematical concept across courses. If a concept already exists (listed below), reuse its exact ID.
{existing_concepts}"""


async def call_openai_chat(
    api_key: str,
    model: str,
    messages: list[dict],
    temperature: float = 0.2,
) -> str:
    async with httpx.AsyncClient(timeout=300) as client:
        resp = await client.post(
            "https://api.z.ai/api/coding/paas/v4/chat/completions",
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {api_key}",
            },
            json={
                "model": model,
                "temperature": temperature,
                "messages": messages,
            },
        )
        resp.raise_for_status()
        envelope = resp.json()
        return envelope["choices"][0]["message"]["content"]


async def call_llm(
    provider: str,
    api_key: str,
    model: str,
    system_prompt: str,
    user_content: str,
) -> str:
    if provider == "anthropic-oauth":
        return await asyncio.to_thread(
            llm_anthropic.call_single_turn, model, system_prompt, user_content, 32000
        )
    if provider == "zai":
        messages = [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_content},
        ]
        return await call_openai_chat(api_key, model, messages, temperature=0.2)
    raise ValueError(f"unknown provider: {provider}")


def parse_json_response(raw: str) -> dict:
    clean = re.sub(r"^```(?:json)?\s*", "", re.sub(r"```\s*$", "", raw.strip()))
    return json.loads(clean)


def build_extend_context(course, all_courses=None) -> str:
    lines = [f"Current course: {course.title}"]
    lines.append(f"Max existing day-id: {max((d.id for d in course.days), default=0)}")
    lines.append("\nExisting phases:")
    for p in course.phases:
        lines.append(f"  {p.num}. {p.title}")
    lines.append("\nExisting days:")
    for d in course.days:
        lines.append(f"  Day {d.id} (phase {d.phase}): {d.title}")
    lines.append("\nExisting concepts in this course:")
    for con in course.concepts:
        lines.append(f"  {con.id} — {con.label}")
    if all_courses:
        other = [(c2.title, c2.concepts) for c2 in all_courses if c2.id != course.id and c2.concepts]
        if other:
            lines.append("\nConcepts from other courses (reuse IDs for same concepts):")
            for title, concepts in other:
                for con in concepts:
                    lines.append(f"  {con.id} — {con.label} (course: {title})")
    return "\n".join(lines)


def build_existing_concepts_block(all_courses) -> str:
    if not all_courses:
        return "(none yet)"
    lines = []
    for c in all_courses:
        for con in c.concepts:
            lines.append(f"  {con.id} — {con.label} (course: {c.title})")
    return "\n".join(lines) if lines else "(none yet)"
